Sources/OutlineView/OutlineViewDragAndDrop.swift

```diff
@@ -0,0 +1,139 @@
+import Foundation
+import AppKit
+
+/// A protocol for use with `OutlineView`, implemented by an object (most likely
+/// the data source) to handle receiving items that are dragged into the `OutlineView`.
+@available(macOS 10.15, *)
+public protocol DropReceiver {
+    associatedtype DataElement: Identifiable
+    
+    /// Given a dragged `NSPasteboardItem`, attempts to read the data in the item
+    /// into a usable object for the `OutlineView`'s data source.
+    ///
+    /// - Parameter item: The Pasteboard item that is being dragged into the OutlineView
+    ///
+    /// - Returns: nil if the Pasteboard item is not readable by this receiver, or
+    ///   a tuple with the decoded item and its associated PasteboardType, which must
+    ///   be included in `acceptedTypes`
+    func readPasteboard(item: NSPasteboardItem) -> DraggedItem<DataElement>?
+    
+    /// Called continuously as a dragged item is moved over the `OutlineView`, this is
+    /// used to define the behavior of the `OutlineView` and the drag cursor.
+    ///
+    /// - Parameter target: A `DropTarget` object describing where the dragged item
+    ///   is being dragged over.
+    ///
+    /// - Returns: A case of `ValidationResult`, which will either highlight the area
+    ///   of the `OutlineView` where the item will be dropped, or some other behavior.
+    ///   See `ValidationResult` for possible return values.
+    func validateDrop(target: DropTarget<DataElement>) -> ValidationResult<DataElement>
+    
+    /// Called once the mouse is released during a drag into the `OutlineView`, after
+    /// `validateDrop(target:)` has returned a case other than `deny`, this function should
+    /// handle updating the data source
+    ///
+    /// - Parameter target: A `DropTarget` item with all the information about
+    ///   where in the `OutlineView` heirarchy the new item will be dropped, and what
+    ///   the item(s) to be dropped is.
+    ///
+    /// - Returns: a boolean indicating that the drop was successful.
+    func acceptDrop(target: DropTarget<DataElement>) -> Bool
+}
+
+@available(macOS 10.15, *)
+public enum NoDropReceiver<Element: Identifiable>: DropReceiver {
+    public typealias DataElement = Element
+    
+    public func readPasteboard(item: NSPasteboardItem) -> DraggedItem<Element>? {
+        fatalError()
+    }
+    
+    public func validateDrop(target: DropTarget<Element>) -> ValidationResult<Element> {
+        fatalError()
+    }
+    
+    public func acceptDrop(target: DropTarget<Element>) -> Bool {
+        fatalError()
+    }
+}
+
+public typealias DragSourceWriter<D> = (D) -> NSPasteboardItem?
+public typealias DraggedItem<D> = (item: D, type: NSPasteboard.PasteboardType)
```

Comment:
Sorry that I didn't ask this before. Why do we need to return the associated pasteboard type for a dragged item? I don't see the example making use of this.

Reply:
Good question :-) I can update the `validateDrop(target:)` function for `OutlineViewDraggingExample.ViewModel` to make use of `type`. It's a small example, but I think important-- when the `DraggedItem` is coming from within the `OutlineView`, you would want to return `.move` and when it's coming from outside, `.copy` to indicate that the result will be added to the data source rather than moved around inside of it.

In general, since `readPasteboard(item:)` can produce `DraggedItem`s from a variety of sources, I wanted to keep the information in there about which `PasteboardType` was used to create the data. I'm sure there could be more uses than just the simple example above.

I'll make that change and commit right now!